### PR TITLE
Add builtins crate

### PR DESCRIPTION
**Pull request by buffalojoec**
#### Problem
The new changes surrounding allocation of minimum required compute units for builtin programs in the cost model have realized a need to track the status of a builtin's migration to BPF. However, currently we are keeping these migration feature gates in two separate lists.

#### Summary of Changes
Introduce a `solana-builtins` crate, which can be the single source of truth for the list of builtins, including their migration feature ID and their CU designation.

### CODE REVIEW

File: builtins/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+        entrypoint: solana_loader_v4_program::Entrypoint::vm,
+        native_cu_cost: solana_loader_v4_program::DEFAULT_COMPUTE_UNITS,
+    }),
+    testable_prototype!(BuiltinPrototype {
+        core_bpf_migration_config: None,
+        name: zk_elgamal_proof_program,
+        enable_feature_id: Some(feature_set::zk_elgamal_proof_program_enabled::id()),
+        program_id: solana_sdk_ids::zk_elgamal_proof_program::id(),
+        entrypoint: solana_zk_elgamal_proof_program::Entrypoint::vm,
+        native_cu_cost: 0, // TODO, these are per-instruction.
```
Comment (buffalojoec) [inline builtins/src/lib.rs]: @tao-stones the two ZK builtins are going to be tricky since they are metered per-instruction.

Comment (tao-stones) [reply builtins/src/lib.rs]: Yea, I don't know what to do with ZK builtins, they are per instruction and they consume much more CUs than other regular builtins. We could consider them as "not builtins" so to allocate 200K by default (Assume users of ZK will set cu-limit sensibly); or we can speed up SIMD-198 to make all remaining builtins per instruction as well (tho that does not address the issue that original builtins consumes hundreds of CUs, while ZK takes thousands of CUs)

Comment (buffalojoec) [reply builtins/src/lib.rs]: Well, you originally didn't have them in the list, so should they not be included here at all? Or is setting them to zero okay?

**Review by LucasSte — DISMISSED** [review]
SVM-wise looks good.

File: builtins/src/lib.rs (hunk 1 above)
Comment (tao-stones) [reply builtins/src/lib.rs]: There was no ZK when the list was created :) I think ZK builtins should be included in the list with "MAX(per instruction cost)"; then:
- before SIMD-170 (eg, current behavior): cost mode reserve enough CUs for all its instructions, maybe a bit over reserving for some, but that is not new. VM's CU Meter always give default 200K for it.
- after SIMD-170, both cost model and CU meter don't care about `native_cu_cost` anymore, they will give 3K CU for it, if user didn't request CU Limit, therefore fails TX.

Comment (tao-stones) [reply builtins/src/lib.rs]: wdyt @jstarry

Comment (jstarry) [reply builtins/src/lib.rs]: I don't really follow what you're suggesting we do in your comment above. But in any case, this PR should definitely not be changing the static list of builtins without a feature gate.

File: builtins-default-costs/src/lib.rs [hunk 2]
```diff
@@ -25,108 +19,35 @@ struct BuiltinCost {
...
 lazy_static! {
-    /// Number of compute units for each built-in programs
-    ///
-    /// DEVELOPER WARNING: This map CANNOT be modified without causing a
-    /// consensus failure because this map is used to calculate the compute
-    /// limit for transactions that don't specify a compute limit themselves as
-    /// of https://github.com/anza-xyz/agave/issues/2212.  It's also used to
-    /// calculate the cost of a transaction which is used in replay to enforce
-    /// block cost limits as of
-    /// https://github.com/solana-labs/solana/issues/29595.
```
Comment (jstarry) [inline builtins-default-costs/src/lib.rs]: Removing the warning isn't going to make this issue go away. We can't update this static map without impacting the  calculated default compute unit limit that we reserve for transactions. Any changes need to be feature gated

**Review by jstarry — CHANGES_REQUESTED** [review]

File: builtins/src/lib.rs (hunk 1 above)
Comment (tao-stones) [reply builtins/src/lib.rs]: Not disagree that the static list shouldn't be changed without feature gate. But technically can still do that before #3799 feature gate is activated.

Comment (jstarry) [reply builtins/src/lib.rs]: Yes we could but I don't think a crate refactoring PR is the right time do that. If we want to update the list before SIMD-0170 is activated we would need to backport the change to v2.0 and v2.1 and update the list of builtin id's included in the SIMD-0170 proposal details.

**Comment by buffalojoec** [comment]
@jstarry @tao-stones @LucasSte I've popped the last commit (where I modify the CU default cost list) from this PR.

I think moving the list of builtins to its own crate is useful for a few reasons. It gives us, and other developers, the ability to access this list, the builtin entrypoint functions, and their respective feature gates (both for enabling and migrating) without having to pull in the whole runtime.

In another follow-up PR, I'll address the cost model problem more directly and see what you guys think.

**Review by jstarry — DISMISSED** [review]

**Review by tao-stones — APPROVED** [review]
lgtm - makes sense to separate default cost related stuff from this PR

**Review by LucasSte — APPROVED** [review]

**Comment by buffalojoec** [comment]
cc @yihau

**Pull request merged**